wandb/sdk/wandb_artifacts.py

```diff
@@ -1443,23 +1443,28 @@ def store_path(
         except self._botocore.exceptions.ClientError as e:
             if e.response["Error"]["Code"] == "404":
                 multi = True
-                start_time = time.time()
-                termlog(
-                    'Generating checksum for up to %i objects with prefix "%s"... '
-                    % (max_objects, key),
-                    newline=False,
-                )
-                objs = (
-                    self._s3.Bucket(bucket)
-                    .objects.filter(Prefix=key)
-                    .limit(max_objects)
-                )
             else:
                 raise CommError(
                     "Unable to connect to S3 (%s): %s"
                     % (e.response["Error"]["Code"], e.response["Error"]["Message"])
                 )
-
+        # S3 doesn't have real folders, however there are cases where the folder key has a valid file which will not
+        # trigger a recursive upload.
+        # we should check the object's metadata says it is a directory and do a multi file upload if it is
+        if "x-directory" in objs[0].content_type:
```

Comment:
this is the content_type of the object when it is a UI created folder: `'application/x-directory; charset=UTF-8'`

Reply:
also, this is a valid key on an object: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Object.content_type